code/controllers/subsystems/processing/odyssey.dm

```diff
@@ -0,0 +1,225 @@
+
+/**
+ * Removes an actor from the subsystem actor list.
+ */
+/datum/controller/subsystem/odyssey/proc/remove_actor(mob/living/carbon/human/H)
+	LAZYREMOVE(actors, H)
+
+/**
+ * Removes a storyteller from the subsystem storyteller list.
+ */
+/datum/controller/subsystem/odyssey/proc/remove_storyteller(mob/abstract/ghost/storyteller/S)
+	LAZYREMOVE(storytellers, S)
+
+/datum/controller/subsystem/odyssey/ui_state()
+	return GLOB.always_state
+
+/datum/controller/subsystem/odyssey/ui_interact(mob/user, datum/tgui/ui)
+	. = ..()
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "OdysseyPanel", "Odyssey Panel")
+		ui.open()
+
+/datum/controller/subsystem/odyssey/ui_data(mob/user)
+	var/list/data = list()
+	if(scenario)
+		data["scenario_name"] = SSodyssey.scenario.name
+		data["scenario_desc"] = SSodyssey.scenario.desc
+		data["scenario_canonicity"] = SSodyssey.scenario.scenario_type == SCENARIO_TYPE_CANON ? "Canon" : "Non-Canon"
+		data["is_storyteller"] = isstoryteller(user)
+
+		if(length(scenario.roles))
+			data["scenario_roles"] = list()
+			for(var/role_singleton in scenario.roles)
+				var/singleton/role/scenario_role = GET_SINGLETON(role_singleton)
+				data["scenario_roles"] += list(
+					list(
+						"name" = scenario_role.name,
+						"desc" = scenario_role.desc,
+						"outfit" = "[scenario_role.outfit]",
+						"type" = scenario_role.type
+					)
+				)
+	return data
+
+/datum/controller/subsystem/odyssey/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+
+	switch(action)
+		if("equip_outfit")
+			if(!ishuman(usr))
+				return
+
+			var/mob/living/carbon/human/player = usr
+			if(player.incapacitated())
+				return
+
+			if(player.z != SSodyssey.scenario_zlevel)
+				to_chat(player, SPAN_WARNING("You can't equip an outfit on a different z-level from the scenario's!"))
+				return
+
+			var/outfit_type = text2path(params["outfit_type"])
+			if(ispath(outfit_type, /obj/outfit))
+				player.delete_inventory(TRUE)
+				player.preEquipOutfit(outfit_type, FALSE)
+				player.equipOutfit(outfit_type, FALSE)
+				return TRUE
+
+		if("edit_scenario_name")
+			if(!isstoryteller(usr))
+				return
+
+			var/new_scenario_name = tgui_input_text(usr, "Insert the new name for your scenario. Remember that this will be visible for anyone in the Stat Panel.", "Odyssey Panel", max_length = MAX_NAME_LEN)
+			if(!new_scenario_name)
+				return
+
+			log_and_message_admins("has changed the scenario name from [SSodyssey.scenario.name] to [new_scenario_name]")
+			SSodyssey.scenario.name = new_scenario_name
+			return TRUE
+
+		if("edit_scenario_desc")
+			if(!isstoryteller(usr))
+				return
+
+			var/new_scenario_desc = tgui_input_text(usr, "Insert the new description for your scenario. This is visible only in the Odyssey Panel.", "Odyssey Panel", max_length = MAX_MESSAGE_LEN)
+			if(!new_scenario_desc)
+				return
+
+			log_and_message_admins("has changed the scenario description")
```

Comment:
Who?

Reply:
this already used usr as an implied argument